pkg/utils/utils.go

```diff
@@ -28,87 +30,166 @@ type RequestData struct {
+	clientIdentityKey, err := walletInstance.GetPublicKey(ctx, opts, "")
 	if err != nil {
-		return nil, errors.New("failed to get client identity key")
+		return nil, fmt.Errorf("failed to get client identity key: %w", err)
+	}
+
+	if clientIdentityKey.PublicKey == nil {
+		return nil, errors.New("client identity key is nil")
 	}
 
 	requestID := generateRandom()
 	encodedRequestID := base64.StdEncoding.EncodeToString(requestID)
 
-	newNonce, err := walletInstance.CreateNonce(context.Background())
+	newNonce, err := sdkUtils.CreateNonce(ctx, walletInstance, wallet.Counterparty{Type: wallet.CounterpartyTypeSelf})
 	if err != nil {
-		return nil, errors.New("failed to create new nonce")
+		return nil, fmt.Errorf("failed to create new nonce: %w", err)
 	}
 
 	var writer bytes.Buffer
 
-	writer.Write(requestID)
+	_, err = writer.Write(requestID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to write request ID: %w", err)
+	}
 
 	request := getOrPrepareTempRequest(requestData)
 	err = WriteRequestData(request, &writer)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to write request data: %w", err)
+	}
+
+	protocol := wallet.Protocol{SecurityLevel: wallet.SecurityLevelEveryAppAndCounterparty, Protocol: auth.AUTH_PROTOCOL_ID}
+
+	baseArgs := wallet.EncryptionArgs{
+		ProtocolID: protocol,
+		Counterparty: wallet.Counterparty{
+			Type:         wallet.CounterpartyTypeOther,
+			Counterparty: serverIdentityKey,
+		},
+		KeyID: fmt.Sprintf("%s %s", newNonce, serverNonce),
+	}
+	createSignatureArgs := wallet.CreateSignatureArgs{
+		EncryptionArgs: baseArgs,
+		Data:           writer.Bytes(),
+	}
+
+	signature, err := walletInstance.CreateSignature(ctx, createSignatureArgs, "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create signature: %w", err)
+	}
+
+	headers := map[string]string{
+		constants.HeaderVersion:     "0.1",
+		constants.HeaderIdentityKey: clientIdentityKey.PublicKey.ToDERHex(),
+		constants.HeaderNonce:       newNonce,
+		constants.HeaderYourNonce:   serverNonce,
+		constants.HeaderRequestID:   encodedRequestID,
+	}
+
+	if signature != nil {
+		headers[constants.HeaderSignature] = hex.EncodeToString(signature.Signature.Serialize())
+	}
+
+	return headers, nil
+}
+
+// PrepareCertificateResponseHeaders prepares the certificate response headers
+func PrepareCertificateResponseHeaders(ctx context.Context, walletInstance interfaces.Wallet, previousResponse *auth.AuthMessage, requestData RequestData) (map[string]string, error) {
+	serverIdentityKey := previousResponse.IdentityKey
+	serverNonce := previousResponse.InitialNonce
+
+	opts := wallet.GetPublicKeyArgs{IdentityKey: true}
+	clientIdentityKey, err := walletInstance.GetPublicKey(ctx, opts, "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get client identity key: %w", err)
 	}
 
-	key, err := ec.PublicKeyFromString(serverIdentityKey)
+	if clientIdentityKey.PublicKey == nil {
+		return nil, errors.New("client identity key is nil")
```

Comment:
```suggestion
		return nil, fmt.Errorf("client identity key is nil")
```
and I think you are using this same error below, so even create separate `var Err...` and use it.